Added JSON Schema (#37)

thzinc opened the pull request:
I'm working on integrating with GBFS feeds and am building a .NET client API to interact with GFBS data. I noticed that JSON Schema files were requested in issue #37, so I figured I'd generate them from the models I created for my client.

For reference, here's the code that I used to generate the JSON schemas: https://github.com/syncromatics/gbfs.net/pull/2

I didn't see a contributors' guide, so please let me know if changes are needed in this PR.

Commit:
Added JSON Schema

afischer commented:
Looks great! Funny—we started working on this year-old request around the same time. I have some [work in my repo](https://github.com/afischer/gbfs/tree/master/schema) but it's still missing a few feeds and I've been trying to find time to finish it.

A few possible suggestions:
- Add `"required"` arrays for arrays/objects—the spec specifies which fields are needed for almost every feed. For example, in the top object, ` "required": ["last_updated", "ttl", "data"]` would make a good addition.
- Minimum and maximum numbers can be added to the schema. This could be helpful for enforcing the 1/0 values, since there is no gbfs "bit" type
- Possibly (?) add descriptions from the spec?

I'm happy to contribute some of my work to your branch if you would like. Thanks for the awesome work!

thzinc commented:
@afischer, I'll see what I can do about each of your suggestions over the next couple of days. Thanks for the feedback!

Commit:
Incorporated PR feedback:

* Specified whether all fields are required or not
* Added min and max values, especially for "boolean" types as defined in the spec
* Added descriptions from the spec

thzinc commented:
@afischer, I've updated this PR with your feedback. I used your work indirectly as reference in building my schema generator for Gbfs.Net.

thzinc commented:
@afischer, or any other maintainer, I was wondering if anyone had a chance to review this PR. If there's someone or a group of someones I should get in contact with, I'd appreciate pointing me in the right direction.

Thanks!

afischer commented:
Unfortunately I'm not a maintainer—just a contributer and former intern at Motivate. I believe @jcn might be the go-to guy for this, but not sure.

thzinc commented:
Ah, thanks @afischer. Sorry for the misattribution. Thanks for your help. :-)

nbdh commented:
This schema definitions have been really helpful during the implementation of GBFS and making sure to return valid data in the correct structure - I wish I had found them/this PR earlier! Thank's a lot for creating them, @thzinc. I wonder why no one cared about the PR as I see no downside in merging them.
As a minor point: I've just both https://github.com/jessedc/ajv-cli and https://github.com/justinrainbow/json-schema to validate with your definitions, both do not accept "Infinity" as maximum for the TTL. I could not find anything about it in the JSON releated specifications. Are you sure that this is a valid expression?

afischer commented:
@nbdh — ~which feed do you have an infinite TTL for? Even for pretty static things like system_information, it seems a little optimistic to say they will never be updated...~ Sorry nbdh, I misread! Yeah, @thzinc that looks like it should be set to something like 1 year (31536000 sec).

@jcn Not sure how involved you are with this still, but I would love to see these merged in! 😄

thzinc commented:
@nbdh and @afischer, thanks for the feedback. I'll try to review the JSON schema for `ttl` today and see what I did wrong.

@nbdh, I'm glad you're finding this helpful!

nbdh commented:
@afischer In my opinion a schema should aim not validate the meaningfulness (or how senseful a certain value is). Even though a TTL of more than a year certainly is not reasonable I'd still consider it valid.
@thzinc I was not implying you did something wrong as I am not an expert on the JSON specs at all. I was just wondering if stating "Infinity" satisfies the specification and if so why both validators I'm using won't interpret it correctly.
I think not stating a maximum would be sufficent?!

thzinc commented:
@nbdh, this JSON Schema is generated from model classes in [gbfs.net](https://github.com/syncromatics/gbfs.net). I think the JSON Schema generator put `infinity` in there, and I'm pretty sure that's invalid.

Commit:
Updated schema to remove invalid "Infinity" from `ttl` range

thzinc commented:
@nbdh, as I suspected, I found that my schema generator was serializing an incorrect value for the `ttl` max value. I've corrected the gbfs.net generator and updated the schema in this PR.

This JSON Schema now reflects the GBFS spec's non-specification of a maximum TTL value. Any integer greater than or equal to zero is valid for `ttl`.

thzinc committed:
Merge remote-tracking branch 'upstream/master'

antrim commented:
Hi! I'm with @MobilityData, pinging on this PR. Between this PR and issue #37, it looks as though there are at least 4 people who support adding a JSON schema definition to the spec -- @afischer, @thzinc, @barbeau, @nbdh. And there aren't any comments against this.

This does add burden for future spec proposals and changes, but that burden seems very small. Some organizations are already finding this is useful as an input to validation tools.

- Are there other data data consumers or validation tools that use this JSON schema?
- Does this JSON schema match the current version of the spec? It looks as though the last update of the PR was Feb 26, 2018 (https://github.com/NABSA/gbfs/pull/70/commits/3976808b18557520771946350f84098a63a91c57), and the last change change to the spec was on June 27, 2018 (f4bfc84b8d6af4f9dfcf9b2e98073fe57e61efef).
- Provided the JSON schema is updated to be current with the spec, should we call a vote on whether to merge this PR?

evansiroky commented:
I'm pinging @PierrickP on this. His [GBFS validator](https://github.com/PierrickP/gbfs-validator) project has [JSON schemas](https://github.com/PierrickP/gbfs-validator/tree/master/gbfs-validator/schema) and I contributed [a change that was accepted](https://github.com/PierrickP/gbfs-validator/pull/3).

yocontra commented:
As a future improvement, I think it would be great to generate the attributes sections of the spec markdown from the JSON schema, instead of having to keep them in sync manually. Will save some time when making changes + ensuring consistency between the two.

ion-biziiac commented on schema/system_hours.json:
```diff
@@ -0,0 +1,102 @@
+          "pattern": "([0-1][0-9]|2[0-3]):[0-5][0-9]:[0-5][0-9]"
+        },
+        "end_time": {
+          "description": "End time for the hours of operation of the system in the time zone indicated in system_information (00:00:00 - 47:59:59). Time can stretch up to one additional day in the future to accommodate situations where, for example, a system was open from 11:30pm - 11pm the next day (i.e. 23:30-47:00)",
```
@thzinc I have a question regarding the `end_time` attribute value. Is it valid and accepted by the GBFS feeds consumers to have the `end_time` stretched up to one additional day in the future? This is the exact situation I am facing now and I just wanted to double check. Thanks.

heidiguenin replied:
It should be. It's explicitly allowed by GBFS' definition of Time generally [Field Types](https://github.com/NABSA/gbfs/blob/master/gbfs.md#field-types):

"Time - Service time in the HH:MM:SS format for the time zone indicated in system_information.json (00:00:00 - 47:59:59). Time can stretch up to one additional day in the future to accommodate situations where, for example, a system was open from 11:30pm - 11pm the next day (i.e. 23:30:00-47:00:00)."

MobilityData will be developing an up-to-date JSON schema and accompanying validator in the coming months, too. @josee-sabourin @mplsmitch

ion-biziiac replied:
@heidiguenin thank you

josee-sabourin commented:
Hi all! 
MobilityData has been working on a schema that we will host on our [own repository](https://github.com/MobilityData/gbfs-json-schema), please let us know what you think either over there in an issue or PR, or in issue #272 in this repo. 

Thanks! 

@thzinc @afischer @nbdh @evansiroky @contra  @ion-biziiac
